docs: initial Elixir SDK documentation (#5530)
dagger/dagger: docs/current/sdk/elixir/756758-get-started.md

- It uses the client's `Dagger.Client.container/2` and `Dagger.Container.from/2` to initialize a new container from the `hexpm/elixir:1.15.4-erlang-25.3.2.5-ubuntu-bionic-20230126` base image.
- It uses `Dagger.Container.with_mounted_directory/3` to mount the project's source files into the container.
- It uses `Dagger.Container.with_exec/3` to define the commands to be executed in the container - in this case, commands such as `mix deps get`, which downloads dependencies and `mix test`, which runs unit tests. Each invocation of `with_exec` returns a revised `Container` with the results of command execution.
- It uses `Dagger.Sync.sync/1` to force command execution.
- It uses `Dagger.close/1` to close the client connection.

Run the Mix task by executing the command below from the project directory:

```shell
dagger run mix a_project.test
```

The `dagger run` command executes the specified command in a Dagger session and displays live progress. Here is an example of the output:

```shell

> Tests succeeded!

Reviewer: @wingyplus following a structure similar to that the Python SDK get started, can you please update your code to display this output at the end?

```

## Step 4: Test against multiple Elixir and Erlang/OTP versions

Now that the Elixir CI tool can test the application against a specified Elixir and Erlang/OTP version, the next step is to extend it for multiple Elixir and Erlang/OTP versions.

Replace the `lib/mix/tasks/a_project.test.ex` file from the previous step with the version below:

```elixir file=snippets/get-started/step4/a_project.test.ex
```

This version has additional support for testing and building against multiple Elixir and Erlang/OTP versions:

- It defines the test matrix, consisting of a list of Elixir and Erlang/OTP version pairs.
- It uses `Task.async_stream/3` to run tests against each version pair concurrently.

Run the tool again by executing the command below:

```shell
dagger run mix a_project.test
```

The tool tests the application, logging its operations to the console as it works. If all tests pass, it displays the final output below:

```shell
Starting tests for Erlang 1.14.5 and Erlang OTP 25.3.2.5
Tests for Python Erlang 1.14.5 and Erlang OTP 25.3.2.5 succeeded!

> TODO

Reviewer: @wingyplus following a structure similar to that the Python SDK get started, can you please update your code to display this type of output at the end? Please then also update this TODO block with the revised output

Author: Updated!

---

> slug: /756758/get-started

Reviewer: ```suggestion
slug: /sdk/elixir/756758/get-started
```

Author: Fixed.

---
slug: /sdk/elixir/756758/get-started
---

# Get Started with the Dagger Elixir SDK

## Introduction

This tutorial teaches you the basic of using Dagger in Elixir. You will learn how to:

- Install Elixir SDK
- Create an Elixir CI task to test an application

> - Improve the Python CI tool to test the application against multiple Elixir and OTP versions

Reviewer: Python?

Author: My bad. Fixed.

## Requirements

This tutorial assumes that:

- You have a basic understanding of the Elixir programming language. If not, [read the Elixir learning](https://elixir-lang.org/learning.html).
- You have a Elixir development environment with Elixir 1.14 or later and Erlang/OTP 25 or later. If not, install [Elixir](https://elixir-lang.org/install.html) and [Erlang/OTP](https://www.erlang.org/downloads).
- You have Docker installed and running on the host system. If not, [install Docker](https://docs.docker.com/engine/install/).

## Step 1: Create an Elixir application

> Create an Elixir project with `mix`:

Reviewer: Should we use `application` or `project`?

Author: We should use project. 🙏

Author: Fixed

```shell

> mix new a_project

Reviewer 2: I would call this project something more descriptive than `a_project`.

What are your thoughts @wingyplus @vikram-dagger on `elixir-with-dagger` or `elixir_with_dagger` instead?

Reviewer: Either of those is good with me

Author: I will be going into the `elixir_with_dagger` way. :)

Then run `mix deps.get` to fetch the Elixir SDK from Hex.pm.

## Step 3: Create the Mix task

Create Elixir module at `lib/mix/tasks/a_project.test.ex`:

```elixir file=snippets/get-started/step3/a_project.test.ex
```

This module performs the following operations:

- It starts all applications that related to Dagger Elixir SDK with `Application.ensure_all_started(:dagger)`.
- It creates a Dagger client with `Dagger.connect!/1`.
- It uses the client's `Dagger.Client.host/1` and `Dagger.Host.directory/3` to address the host directory. Uses `exclude` in `Dagger.Host.directory/3` to filter unwanted files and directories.
- It uses the client's `Dagger.Client.container/2` and `Dagger.Container.from/2` to initialize a new container and uses `hexpm/elixir:1.15.4-erlang-25.3.2.5-ubuntu-bionic-20230126` as a base image.
- It uses `Dagger.Container.with_mounted_directory/3` to mount source files into container.
- It uses `Dagger.Container.with_exec/3` to execute a command.
- It uses `Dagger.Sync.sync/1` to force executing commands.
- It uses `Dagger.close/1` to teardown the client.

Run the Mix task by executing the command below from the project directory:

```shell
mix a_project.test

> ```

Reviewer 2: What are your thoughts about adding `dagger run` in front of this command @wingyplus? The output will look a lot better! FWIW, it will require downloading & installing the `dagger` CLI: https://docs.dagger.io/cli/465058/install

This output will become especially important when testing against multiple versions below.

Reviewer: @wingyplus in my copy edit I have modified these commands to use `dagger run`, as this is now the recommended approach. Please double-check that the output is as expected.

Author: I update the output here but please note that we have https://github.com/dagger/dagger/issues/5666, it stuck when running with `dagger run` command.

Reviewer 2: Works on both macOS & NixOS for me: https://github.com/dagger/dagger/issues/5666#issuecomment-1740760580

## Step 4: Test against multiple Elixir and Erlang/OTP versions

Now that the Elixir CI tool can test the application against a Elixir and Erlang/OTP version, the next step is to extend it for multiple Elixir and Erlang/OTP versions.

Replace the `lib/mix/tasks/a_project.test.ex` file from the previous step with the version below:

```elixir file=snippets/get-started/step4/a_project.test.ex
```

> In this version has additional support for test and building against multiple Elixir and Erlang/OTP versions.

Reviewer: ```suggestion
This version has additional support for testing and building against multiple Elixir and Erlang/OTP versions.
```

> - It defines the test matrix, by defining a list of Elixir and Erlang/OTP version pair.

Reviewer: ```suggestion
- It defines the test matrix, by defining a list of Elixir and Erlang/OTP version pairs.
```

- It pass through the `Task.async_strem/3` to run test in each version concurrently.

## Conclusion

> This turorial introduced you to the Dagger Elixir SDK. It explaned how to install the SDK and use it with an Elixir application.  It also provided a working example of the a tool powered by the SDK, demonstrating how to test an application against multiple Elixir and Erlang/OTP versions in parallel.

Reviewer: ```suggestion
This tutorial introduced you to the Dagger Elixir SDK. It explaned how to install the SDK and use it with an Elixir application.  It also provided a working example of a CI tool powered by the SDK, demonstrating how to test an application against multiple Elixir and Erlang/OTP versions in parallel.
```

> Use the [HexDocs SDK Reference](https://hexdocs.pm/dagger/Dagger.html) to learn more abount the Dagger Elixir SDK.

Reviewer: ```suggestion
Use the [HexDocs SDK Reference](https://hexdocs.pm/dagger/Dagger.html) to learn more about the Dagger Elixir SDK.
```
